### DISCUSSION TITLE
111 Boot's - Software/System restart in 4 days

### DISCUSSION THREAD

**Discussion by martin072** [arendst/Tasmota#16222] [Q&A] [answered]
I am having trouble with one of my tasmata's (a Wemos D1 Mini with a WS2812 strip attached), it restarts very often, about 11 times in 2 days, however, after resetting the boot count, it shows 111 boots in just 2 days?? Weird thing is, the reason is **Software/System restart ** and not a crash. 

I've enable debug-syslog, but it shows nothing strange. Last messages are either 'ESP-WIF: Sending Gratuitous ARP' -or 'ESP-WIF: Checking connection...' and then a reset happens around 20~30 seconds after that last message. No crash or something (at least not able to check that..) 

These Restarts happen when the lights are off (sometimes when they are on though, but last two days I was not at home, lights were off all the time, still got the resets). Not sure if it could be power supply related?

Any idea's?

**Comment by sfromis** [comment]
Depending on details of your power source, that is always a potential risk factor. Beware that many D1 Mini clones have a paltry voltage regulator (~150 mA 3.3V, instead of 500 mA) not great when adding stuff to the board.

Remote logging via syslog etc may not show much for some types of trouble. Hooking up to serial and monitoring may provide more.

**Reply by joba-1** [reply]
what is the base lib tasmota uses for ws2812?

I ask because I had that problem for years with my own firmware [NeoXmas](https://github.com/joba-1/NeoXmas).
Finding the root cause failed - no messages via serial when it happened. No difference if leds were on or off. Big power supply, different ESP boards. 
Recently I changed from Adafruit_NeoPixels to NeoPixelBus and all problems gone. Porting effort was marginal.

**Reply by sfromis** [reply]
Tasmota uses https://github.com/arendst/Tasmota/tree/development/lib/lib_basic/NeoPixelBus too

You have no real basis for saying "*that*" problem, just that you also have had some problem.

**Reply by joba-1** [reply]
> You have no real basis for saying "_that_" problem, just that you also have had some problem.

oh well, isn't that a bit picky? :) My basis was that the frequency and every other detail martin described was identical: Same problem, but unfortunately different cause. Now I'm out of ideas. Except maybe signal quality issues?

**Reply by sfromis** [reply]
Unexpected boots is so imprecise that it is not picky to not just jump into seeing it as *same* problem.

**Comment by martin072** [comment]
Might have found a lead towards the resets... 

My Tasmota is running a relative big script to control the led strep (Its a custom script controlling 1 strip divided in three parts, and controlling those 3 segments separately.) I've disabled the script and the resets stopped. 

Would there be any way to debug the script?

**Reply by pkkrusty** [reply]
Stick a bunch of `print Now In Section X` into your script, that'll help you narrow it down. Maybe some `delay(x)` too so you have time to see it in the console before the system crashes.

**Reply by barbudor** [reply]
You might also monitor the free memory in the information page or through the STATE message published at Teleperiod ("heap")

**Reply by martin072** [reply]
> Stick a bunch of `print Now In Section X` into your script, that'll help you narrow it down. Maybe some `delay(x)` too so you have time to see it in the console before the system crashes.

That would be impossible, as the loops are in the 100ms and 1s so it will flood the console, and I have also no idea when it resets, so I would need to stare at the console for hours and hours..



> You might also monitor the free memory in the information page or through the STATE message published at Teleperiod ("heap")

It varies between 8K and 13K in between resets. Resets seem random, longest uptime was 5hrs, lowest was 30min.
Last STATE message before the reset, HEAP varies between 8k and 13k

Removed the >W section to test something, also removed the variables that were used in >W, funny thing it restarted when I wanted to Save the Script and did 2 boots (no exception recorded by tasmota). Second save worked.. 
Using Tasmota 12.0.2 (custom compile), but version 11.x had same issue.

This is enabled for my scripting:

```
  #define USE_BUTTON_EVENT
    #define USE_SCRIPT_SUB_COMMAND
    #define USE_SCRIPT_STATUS
    #define SUPPORT_MQTT_EVENT
    #define USE_SCRIPT_WEB_DISPLAY
    #define SCRIPT_FULL_WEBPAGE
    #define SCRIPT_STRIP_COMMENTS
    #define USE_SCRIPT_GLOBVARS
    #define USE_SCRIPT_TIMER
    #define SCRIPT_POWER_SECTION                   // enable >P section (execute on power changes)
    #define LARGE_ARRAYS
```

**Reply by sfromis** [reply]
> It varies between 8K and 13K in between resets

8K is too low, and 13K is also lower than I'd want to go. You may have to look into cutting out memory-hungry features. Maybe consider switching to ESP32 with more headroom, and better programming features in Tasmota.

**Reply by martin072** [reply]
What would ideal? I have basically made 1 binary for all my projects, so it's easier to update them. 
The ones with scripting enabled average between 12 - 15K of free memory, is that still to low? I have one that has no scripting enable because it doesn't use it, and its an old esp8266 1M module, that one shows 22K of free memory. Obviously I would like to keep the scripting features ;)

**Reply by sfromis** [reply]
No hard number, but my view is that over 20K is just fine, after that I'd like to cut down on usage. About 12K is a bare minimum. You do get a lot more available with ESP32, from a couple of random board: `"Heap":188` or `"Heap":108`

**Reply by martin072** [reply]
Ok, will see if I can cut some stuff out. Not so easy to change existing 8266 to esp32 for me, but thanks for the info.

**Reply by sfromis** [reply]
You can get an ESP32-based board which is basically pin-compatible with the D1 Mini, like the Wemos S2 Mini. Easy swap.  Maybe prefer genuine Wemos to lookalikes, at least one user reported trouble with an unbranded clone

**Reply by martin072** [reply]
So I've  noticed. Will have to look, believe they are slightly bigger, needs to fit the pcb's. Thanks again.

**Reply by sfromis** [reply]
Just compared mine, *exactly* the same size. The extra rows of pins which you do not have to use for anything is on the inside of the D1 mini matching compatible rows.

BTW, the similar Wemos C3 Mini has proven to have a bad antenna design, don't know if they've done anything to fix that. My Wemos S2 Mini boards have worked without issue.

**Reply by martin072** [reply]
Yes, just looked at it, however, the ESP32_Mini is slightly offset due to the pins that are not used, therefore it overlaps a few terminal blocks on my PCB. So it's not an exact fit, but I think it will work, just need to remove the ESP32 if I need to screw/unscrew from the terminal blocks. 

Will give it a go if I can make some time for it. Will let you know. 

Mean time I've removed all that I could on the ESP8266 build, still it only gives me just 12/13k of heap. Will monitor

**Reply by martin072** [reply]
Ahh.. Wait, now I see what you mean with the S2.. I have a different one: [https://nl.aliexpress.com/item/1005001621844145.html?spm=a2g0o.order_list.0.0.21ef79d2pFJ2nh&gatewayAdapt=glo2nld](ESP32 D1 Mini)

That is slightly bigger. Will order some S2's to see how that works. Thanks!

**Reply by sfromis** [reply]
If you do not solder anything to the extra rows, they take no space at all. Assuming MCU+antenna side on top, the S2 mini even has less risk of conflicts, as it only has components (including USB socket) on one side. And I see no offsets when having both boards in my hands.

I suppose that USB+buttons on the top side of the board could be an "inconvenient" difference in case of a fit to a custom 3D printed box.

**Reply by sfromis** [reply]
And that unbranded `D1 Mini ESP32 ESP-32` does not appear to be that pin-compatible with the original ESP8266-based Wemos D1 Mini anyway, things extra pin counts in all the rows, and a much wider+longer board too.

**Comment by martin072** [comment, marked as answer]
So far its running for 26hrs, no resets. Only thing I've done is commented out some unused drivers and unused script features and removed the >W section including the string variables. 

Hard to say what caused it then, Heap is still only 12.2 ..

**Reply by martin072** [reply]
Will close this, after removing some features I don't need, it seems to be very stable. running for 7 days now with 0 restarts.
